### PR TITLE
ANN: Detect E0120: impl Drop for Trait is illegal

### PR DESCRIPTION
**Pull request by mibac138**
Related to #886

### CODE REVIEW

File: src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt [hunk 1]
```diff
@@ -297,6 +298,13 @@ class RsErrorAnnotator : RsAnnotatorBase(), HighlightRangeExtension {
         }
     }
 
+    private fun checkImplDropForTrait(holder: AnnotationHolder, impl: RsImplItem, traitRef: RsTraitRef, trait: RsTraitItem) {
+        if (impl.typeReference?.baseType?.path?.reference?.resolve() !is RsTraitItem) return
+        if (trait.qualifiedName == "core::ops::drop::Drop") {
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt]: See https://github.com/intellij-rust/intellij-rust/pull/3789#discussion_r281040984

File: src/test/kotlin/org/rust/ide/annotator/RsStdlibErrorAnnotatorTest.kt [hunk 2]
```diff
@@ -24,4 +24,11 @@ class RsStdlibErrorAnnotatorTest : RsAnnotatorTestBase(RsErrorAnnotator::class.j
 
         <error descr="Can't find crate for `litarvan` [E0463]">extern crate litarvan;</error>
     """)
+
+    fun `test impl Drop for Trait E0120`() = checkErrors("""
```
Comment (Undin) [inline src/test/kotlin/org/rust/ide/annotator/RsStdlibErrorAnnotatorTest.kt]: See https://github.com/intellij-rust/intellij-rust/pull/3789#discussion_r281042654 to move this test to `RsErrorAnnotatorTest.kt`

**Comment by mibac138** [comment]
@Undin I've addressed your comments

**Comment by Undin** [comment]
@mibac138 This branch, as well as #3789, has conflicts with master 
But looks like it will have conflicts with master when #3789 will be merged even you rebase onto master now
So, let's rebase this branch onto master only when #3789 will be merged to avoid additional work

**Comment by mibac138** [comment]
Rebased onto #3789. I think that should avoid creating conflicts once #3789 is merged?

**Comment by vlad20012** [comment]
Hmm, your check is unreasonably very specific. `Drop` can be implemented only for structs and enums, so E0120 can be also showed when implementing `Drop` for reference, pointers, arrays, primitives, etc. Looks like it's easy to detect attempt to impl `Drop` for NOT struct or enum and show the error in this case

**Comment by mibac138** [comment]
@vlad20012 That's a fair point, I must admit I got carried away by the trait part and forgot about other types. I have fixed that now and added tests for the types that have been added.

File: src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt [hunk 3]
```diff
@@ -353,6 +351,15 @@ class RsErrorAnnotator : RsAnnotatorBase(), HighlightRangeExtension {
         }
     }
 
+    // E0120: Drop can be only implemented by structs and enums
+    private fun checkImplDropForNonAdtError(holder: AnnotationHolder, impl: RsImplItem, traitRef: RsTraitRef, trait: RsTraitItem) {
+        if (trait != trait.knownItems.Drop) return
+
+        if (impl.typeReference?.type is TyAdt) return
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt]: Looks dangerous for the case of Drop impl with missed type reference, i.e.
```rust
impl Drop for {}
```

Let's add a test that show that we don't have an error in this case.

**Comment by mibac138** [comment]
Added more tests

**Comment by vlad20012** [comment]
:+1: 
bors r+

**Pull request merged**